sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala

```diff
@@ -0,0 +1,133 @@
+    val input = InternalRow(1)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """{"a":1}""")
+  }
+
+  test("initial with StructType and write out rows") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = new GenericArrayData(InternalRow(1) :: InternalRow(2) :: Nil)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """[{"a":1},{"a":2}]""")
+  }
+
+  test("initial with StructType and write out an array with single empty row") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = new GenericArrayData(InternalRow(null) :: Nil)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """[{}]""")
+  }
+
+  test("initial with StructType and write out an empty array") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = new GenericArrayData(Nil)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """[]""")
+  }
+
+  test("initial with Map and write out a map data") {
+    val dataType = MapType(StringType, IntegerType)
+    val input = ArrayBasedMapData(Map("a" -> 1))
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """{"a":1}""")
+  }
+
+  test("initial with Map and write out an array of maps") {
+    val dataType = MapType(StringType, IntegerType)
+    val input = new GenericArrayData(
+      ArrayBasedMapData(Map("a" -> 1)) :: ArrayBasedMapData(Map("b" -> 2)) :: Nil)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """[{"a":1},{"b":2}]""")
+  }
+
+  test("error handling: initial with StructType but error calling write a map") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = ArrayBasedMapData(Map("a" -> 1))
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    intercept[Exception] {
+      gen.write(input)
+    }
+  }
+
+  test("error handling: initial with StructType but error calling write an array of maps") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = new GenericArrayData(
+      ArrayBasedMapData(Map("a" -> 1)) :: ArrayBasedMapData(Map("b" -> 2)) :: Nil)
+    val writer = new CharArrayWriter()
+    val gen = new JacksonGenerator(dataType, writer, option)
+    intercept[Exception] {
```

Comment:
This looks throwing `ClassCastException` rather than `UnsupportedOperationException` as I guess intended. 
One workaround I could think now is, calling it by named arguments, for example:

```diff
- writeObject(writeMapData(map, dataType.asInstanceOf[MapType]), mapElementWriter)
+ writeObject(writeMapData(
+       fieldWriter = mapElementWriter,
+       map = map,
+       mapType = dataType.asInstanceOf[MapType]))
```

Reply:
And also, we should catch narrower exception type if possible, `intercept[UnsupportedOperationException] {` - [scala-style-guide#intercepting-exceptions](https://github.com/databricks/scala-style-guide#intercepting-exceptions).

Reply:
Oh, this should be a workaround to try. Then we can get `UnsupportedOperationException`.

Reply:
This way is great!  Thanks :)

Reply:
@HyukjinKwon @viirya  
umm... I still have problems of `ArrayType[MapType]` and `ArrayType[StructType]`. We are so hard to check which type is the elements of `ArrayData` that we can't throw `UnsupportedOperationException`. If we need to catch a narrower exception, maybe we can just catch `ClassCastException`?
Just like `interecept[ClassCastException] {`

Reply:
I see. I think this array type validation was already failed though. Let's remove out the test cases related with `ClassCastException` too and let's do this in a separate PR. I think this does not block this PR as it is not a regression.

Reply:
ok. I'll remove them.

Reply:
Actually since this is for internal use, and I assume the callers know what they are doing and provide correct arrays. But we can add a TODO comment for it.